xapian-core/cluster/kmeans.cc

```diff
@@ -0,0 +1,142 @@
+ * published by the Free Software Foundation; either version 2 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301
+ * USA
+ */
+
+#include <config.h>
+
+#include "xapian/cluster.h"
+#include <debuglog.h>
+#include <xapian/error.h>
+
+#include <cstdlib>
+#include <climits>
+#include <set>
+#include <unordered_map>
+#include <vector>
+
+#define EPSILON 0.0000000001
+
+using namespace Xapian;
+using namespace std;
+
+string
+KMeans::get_description() const {
+    LOGCALL(API, string, "KMeans::get_description()", NO_ARGS);
+    return "KMeans clusterer";
+}
+
+void
+KMeans::initialize_centroids(ClusterSet &cset) {
+    LOGCALL_VOID(API, "KMeans::initialize_random()", cset);
+    set<docid> centroid_docs;
+    int size = docs.size();
+    srand(time(NULL));
+    for (unsigned int i = 0; i < k;) {
+	int x = rand() % size;
+	if (centroid_docs.find(x) == centroid_docs.end()) {
+	    Point a = docs[x];
+	    Centroid centroid;
+	    centroid.set_to_point(a);
+	    centroids.push_back(centroid);
+	    centroid_docs.insert(x);
+	    Cluster cluster1(centroid);
+	    cset.add_cluster(cluster1);
+	    ++i;
+	}
+    }
+}
+
+void
+KMeans::initialize_points(const MSet &source, TermListGroup &tlg) {
+    LOGCALL_VOID(API, "KMeans::initialize_points()", source | tlg);
+    for (MSetIterator it = source.begin(); it != source.end(); ++it) {
+	Point p;
+	p.initialize(tlg, it.get_document());
+	docs.push_back(p);
+    }
+}
+
+bool
+KMeans::converge(vector<Centroid> &previous, vector<Centroid> &current) const {
+    LOGCALL(API, bool, "KMeans::converge()", previous | current);
+    CosineDistance cosine;
+    for (unsigned int i = 0; i < k; ++i) {
+	Centroid &prev = previous[i];
+	Centroid &curr = current[i];
+	double dist = cosine.similarity(prev, curr);
+	if (dist > EPSILON)
+	    return false;
+    }
+    return true;
+}
+
+ClusterSet
+KMeans::cluster(MSet &mset) {
+    LOGCALL(API, ClusterSet, "KMeans::cluster()", mset);
+    unsigned int size = mset.size();
+    mset.fetch();
+    if (size <= 0)
+	throw AssertionError("Size of MSet should be greater than zero");
+    if (k <= 0)
+	throw AssertionError("Number of clusters should be greater than zero");
+    if (k > size)
+	throw AssertionError("The number of clusters cannot be greater than number of documents in MSet");
+    ClusterSet cset;
+    TermListGroup tlg;
+    tlg.add_documents(mset);
+    CosineDistance cosine;
+    initialize_points(mset, tlg);
+    initialize_centroids(cset);
+    double min = INT_MAX;
+    double min_cluster = 0;
```

Comment:
It might be helpful to make these variable names more expressive. Perhaps `closest_cluster_number` and `distance_to_closest_cluster` or something.